### ISSUE TITLE
name provider is a lot slower than e.g. word provider

### ISSUE THREAD

**Issue by mfussenegger** [joke2k/faker#421]
Not sure if this is of interested, but I've noticed that the name provider is significantly slower than other providers. Which seems kind of odd:

```
from faker import Factory
import timeit

fake = Factory.create()
print(timeit.timeit('fake.word()', globals=locals()))
print(timeit.timeit('fake.name()', globals=locals()))
```

```
python faker-perf.py
2.0924576180004806
611.8756651770018
```

**Comment by jchildren** [comment]
I might be wrong, but I think this is because `faker.utils.distribution.choice_provider` takes linear time due to comprehensions for the cumulative distribution function used to normalize probabilities. If probabilities were normalized in advance (or assumed to be normalised), it would be significantly faster but I can see the advantage in not assuming this for flexibility.

It might be possible to work out the cdfs for a generator when the provider is added, but this might get complicated.

**Comment by jchildren** [comment]
As a point of interest, I tried out using `random.choices` from python 3.6 in place of `choice_distribution` and found that it was a few times faster with the same benchmark. Obviously this is to be expected given that `random.choices` is implemented in c, but is it worth adding the option to use `random.choices` instead if running python 3.6?

with random.choices
```
>>> print(timeit.timeit('fake.name()', globals=locals()))
111.96085176300039
```
without
```
>>> print(timeit.timeit('fake.name()', globals=locals()))
458.3832480560013
```

I'm not sure what the best way to support this option is other than using hasattr on the instance of random.

**Comment by catleeball** [comment]
@jchildren an alternative to hasattr might be using sys.version_info. You could do something like:

```python
from sys import version_info

if version_info.major >= 3 and version_info.minor >= 6:
     from random import choices
    # use choices
else:
    # use choice_distribution
```

There might be a cleaner way, but I've used this solution in the past and it worked fine.

**Comment by catleeball** [comment]
Started dinking with this a bit ago, tried something like this:

```python
def choice_distribution(a, p):
    assert len(a) == len(p)

    if version_info.major >= 3 and version_info.minor >= 6:
        from random import choices
        return choices(a, cum_weights=p)[0]
    else:
        cdf = list(cumsum(p))
        normal = cdf[-1]
        cdf2 = [float(i) / float(normal) for i in cdf]
        uniform_sample = random_sample()
        idx = bisect.bisect_right(cdf2, uniform_sample)
        return a[idx]
```

But it gets caught on this unit test:

```python
self.assertTrue(boundaries[0][0] > a_pop > boundaries[0][1])
```

I'll probably keep playing with this later. I think I might have misunderstood how the current choice_distribution works, will have to read it more closely when I have a minute. Let me know if you spot whatever silly mistake I'm making. :)

**Comment by jchildren** [comment]
@catleeball it seems you just need to change `cum_weights` to `weights` and the test passes. This logic was done by `cdf = list(cumsum(p))` in the current implementation, but the `choices` function can do this for us!

I think this could probably be submitted as a pull request now?

**Comment by catleeball** [comment]
Ah, thanks for the tip @jchildren ! I just made PR #553 for this bug.

**Comment by jchildren** [comment]
After trying it again it seems that the test seems to still occasionally fail, so that doesn't fix the problem but I think `weights` is still correct.

**Comment by catleeball** [comment]
Weirdly, switching to using `hasattr()` made unit tests and the pipeline happy, and it's faster to boot.

**Comment by malefice** [comment]
@fcurella This appears to have been resolved. My PR #1037 actually built on top of the work done here. Are there still outstanding items that need to be addressed, or can this issue be closed?

**Comment by fcurella** [comment]
Thanks for pointing this out @malefice ! This issue can be closed/

Closed by #553
